libpod/options.go

```diff
@@ -1339,17 +1340,46 @@ func WithCommand(command []string) CtrCreateOption {
 // This creates a container from a directory on disk and not an image.
 func WithRootFS(rootfs string) CtrCreateOption {
 	return func(ctr *Container) error {
+		var err error
 		if ctr.valid {
 			return define.ErrCtrFinalized
 		}
-		if _, err := os.Stat(rootfs); err != nil {
+
+		rootfs, err = withRootFSOverlay(ctr, rootfs)
+		if err != nil {
+			return err
+		}
+		if _, err = os.Stat(rootfs); err != nil {
 			return err
 		}
+
 		ctr.config.Rootfs = rootfs
 		return nil
 	}
 }
 
+func withRootFSOverlay(ctr *Container, rootfs string) (string, error) {
+	// Note: This is just an experiment and will be more generic in final draft.
+	parts := strings.SplitN(rootfs, ":", 2)
+	// lookout for --rootfs /some/path:O
+	if len(parts) > 1 && parts[1] == "O" {
+		overlaySrc := parts[0]
+		// remove tmp from here and take value of --tmpfs from cli
+		overlayDest := ctr.runtime.store.GraphRoot()
+		contentDir, err := overlay.GenerateStructure(ctr.runtime.store.GraphRoot(), ctr.ID(), "rootfs", ctr.RootUID(), ctr.RootGID())
+		if err != nil {
+			return "", errors.Wrapf(err, "rootfs-overlay: failed to create TempDir in the %s directory", overlayDest)
+		}
+		overlayMount, err := overlay.Mount(contentDir, overlaySrc, overlayDest, ctr.RootUID(), ctr.RootGID(), ctr.runtime.store.GraphOptions())
+		if err != nil {
+			return "", errors.Wrapf(err, "rootfs-overlay: creating overlay failed %q", overlaySrc)
+		}
+
+		return overlayMount.Source, nil
+	}
+	return rootfs, nil
```

Comment:
would it make sense to reverse the conditional logic to avoid a return the if condition?